include/unifex/indexed_for.hpp

```diff
@@ -0,0 +1,169 @@
+  UNIFEX_NO_UNIQUE_ADDRESS Predecessor pred_;
+  UNIFEX_NO_UNIQUE_ADDRESS Policy policy_;
+  UNIFEX_NO_UNIQUE_ADDRESS Range range_;
+  UNIFEX_NO_UNIQUE_ADDRESS Func func_;
+
+  template <template <typename...> class Tuple>
+  struct indexed_for_result {
+   public:
+    template <typename... Args>
+    using apply = Tuple<Args...>;
+  };
+
+  template <typename... Args>
+  using is_overload_noexcept = std::bool_constant<noexcept(
+      std::invoke(std::declval<Func>(), std::declval<Args>()...))>;
+
+  template <template <typename...> class Variant>
+  struct calculate_errors {
+   public:
+    template <typename... Errors>
+    using apply = std::conditional_t<
+        Predecessor::
+            template value_types<std::conjunction, is_overload_noexcept>::value,
+        Variant<Errors...>,
+        deduplicate_t<Variant<Errors..., std::exception_ptr>>>;
+  };
+
+  template <
+      template <typename...> class Variant,
+      template <typename...> class Tuple>
+  using value_types = deduplicate_t<typename Predecessor::template value_types<
+      Variant,
+      indexed_for_result<Tuple>::template apply>>;
+
+  template <template <typename...> class Variant>
+  using error_types = typename Predecessor::template error_types<
+      calculate_errors<Variant>::template apply>;
+
+  friend constexpr auto tag_invoke(
+      tag_t<blocking>,
+      const indexed_for_sender& sender) {
+    return blocking(sender.pred_);
+  }
+
+  template <typename Receiver>
+  struct indexed_for_receiver {
+    UNIFEX_NO_UNIQUE_ADDRESS Func func_;
+    UNIFEX_NO_UNIQUE_ADDRESS Policy policy_;
+    UNIFEX_NO_UNIQUE_ADDRESS Range range_;
+    UNIFEX_NO_UNIQUE_ADDRESS Receiver receiver_;
+
+    // sequenced_policy version supports forward range
+    template<typename... Values>
+    void apply_func_with_policy(const execution::sequenced_policy& policy, Range&& range, Func&& func, Values&... values)
+        noexcept(noexcept(std::invoke((Func &&) func_, std::declval<typename Range::iterator::value_type>(), values...))) {
+      for(auto idx : range) {
+        std::invoke((Func &&) func, idx, values...);
+      }
+    }
+
+    // parallel_policy version requires random access range
+    template<typename... Values>
+    void apply_func_with_policy(const execution::parallel_policy& policy, Range&& range, Func&& func, Values&... values)
+        noexcept(noexcept(std::invoke((Func &&) func_, std::declval<typename Range::iterator::value_type>(), values...))) {
+      auto start = range.begin();
+      for(auto idx = 0; idx < range.size(); ++idx) {
+        std::invoke((Func &&) func, start[idx], values...);
+      }
+    }
+
+    template <typename... Values>
+    void set_value(Values&&... values) && noexcept {
+      if constexpr (noexcept(std::invoke(
+                        (Func &&) func_, std::declval<typename Range::iterator::value_type>(), values...))) {
+        apply_func_with_policy(policy_, (Range&&) range_, (Func &&) func_, values...);
```

Comment:
This implementation is passing through the `values` passed to `set_value()` directly to the function.
This is fine because the invocation of `func_` is blocking and is guaranteed to complete before `set_value()` returns.

But what if some implementation wants to execute these concurrently on multiple threads?
We may not want to block the call to `set_value()` on the completion of tasks on other threads and so would presumably need to copy the `values...` to a stable storage location (in the operation-state?) first before launching the tasks on other threads.

Those specialisations may have to alter the `value_types` to apply `remove_cvref_t<Values...>` as they will be forwarding through the copies when it completes, rather than the original values passed to `set_value()`.